File: parallel-consumer-proxy-clients/parallel-consumer-proxy-client-ruby/demo/lib/parallel_consumer/demo/sidecar.rb
# Copyright (C) 2026 Antony Stubbs and contributors
# frozen_string_literal: true

module ParallelConsumer
  module Demo
    # Where the sidecar binary is, for a language that cannot build one.
    #
    # THE SIDECAR IS A JVM PROGRAM AND RUBY IS NOT A JVM LANGUAGE, which is the whole difficulty
    # this class holds. The Java demo asks its own runtime - `System.getProperty("java.class.path")`
    # is already the answer there. Ruby has nothing to ask, so the launcher and the classpath are
    # handed in by whichever entry point ran: `demo/run.sh` computes them with Maven natively, and
    # the demo image bakes them in at build time.
    #
    # THAT IS WHY THESE TWO VARIABLES ARE NOT PART OF THE FLAG CONTRACT. Every `--flag` in
    # `parallel-consumer-proxy/demo/README.md` has a `PC_DEMO_` variable and this pair has no flag:
    # they are plumbing between the entry point and the demo, not a dial a reader turns. A flag
    # would invite pointing the demo at an arbitrary binary, and which binary receives your Kafka
    # credentials is a security decision the client library deliberately refuses to make (its
    # `SidecarCommand` takes an absolute path and never a PATH lookup).
    #
    # THE COMMAND IS `java` ITSELF, NEVER A WRAPPER SCRIPT. Parent death is detected by EOF on the
    # inherited pipe (KTD19); a shell wrapper inherits the write end and holds it open, which
    # defeats the signal and leaks a JVM that still holds Kafka group membership.
    class Sidecar
      MAIN_CLASS = "bz.stub.parallelconsumer.proxy.Main"

      CLASSPATH_VARIABLE = "PC_DEMO_SIDECAR_CLASSPATH"

      JAVA_VARIABLE = "PC_DEMO_SIDECAR_JAVA"

      # `.freeze` because the message interpolates a constant, so `frozen_string_literal` does not
      # cover it - RuboCop's Style/MutableConstant is what noticed.
      MISSING = <<~MESSAGE.freeze
        #{CLASSPATH_VARIABLE} is not set, so this demo does not know where the sidecar is.

        Start the demo through its own entry point, which builds the sidecar and sets it:

            parallel-consumer-proxy-clients/parallel-consumer-proxy-client-ruby/demo/run.sh
      MESSAGE

      attr_reader :path, :args

      def initialize(path, args)
        @path = path
        @args = args
      end

      # The command that runs the sidecar, or a raise saying how to get one.
      def self.resolve(env = ENV)
        classpath = env[CLASSPATH_VARIABLE].to_s
        raise MISSING if classpath.strip.empty?

        new(java_binary(env), ["-cp", classpath, MAIN_CLASS])
      end

      # The JVM launcher, as an ABSOLUTE path - the client library refuses anything else, and a
      # relative or PATH-resolved launcher is exactly the ambiguity it refuses it for.
      #
      # A PATH SEARCH IS ACCEPTABLE HERE AND NOWHERE ELSE: this is a demo choosing a JVM to run a
      # sidecar it just built, not a library choosing which binary receives a user's credentials.
      def self.java_binary(env)
        candidates = [env[JAVA_VARIABLE], env["JAVA_HOME"] && File.join(env["JAVA_HOME"], "bin", "java")]
        candidates += env.fetch("PATH", "").split(File::PATH_SEPARATOR).map { |dir| File.join(dir, "java") }
        found = candidates.compact.find { |candidate| File.executable?(candidate) }
        raise "no JVM found - set #{JAVA_VARIABLE} or JAVA_HOME" unless found

        File.expand_path(found)
      end
      private_class_method :java_binary
    end
  end
end

File: parallel-consumer-proxy-clients/parallel-consumer-proxy-client-go/demo/sidecar.go
// Copyright (C) 2026 Antony Stubbs and contributors

package main

import (
	"errors"
	"fmt"
	"os"
	"os/exec"
	"path/filepath"
	"strings"
)

// sidecarClasspathEnv names the file holding the sidecar's classpath.
//
// IT IS PLUMBING, NOT AN EIGHTH FLAG. The contract's environment rule is that every FLAG has a
// PC_DEMO_ variable; this one has no flag and no default a user would set, because it exists only
// because the shipped sidecar is not yet a binary. run.sh writes the file and sets this; the
// Dockerfile bakes both in. demo/README.md records it as a Go divergence for the integrator.
const sidecarClasspathEnv = "PC_DEMO_SIDECAR_CLASSPATH"

// mainClass is the sidecar's entry point: bz.stub.parallelconsumer.proxy.Main, which binds an
// ephemeral loopback port, prints "port: <n>", and serves until its parent dies.
const mainClass = "bz.stub.parallelconsumer.proxy.Main"

// sidecarCommand is what the client library spawns: an ABSOLUTE binary path plus its arguments.
//
// TODAY THAT BINARY IS THE JVM LAUNCHER AND THE PROXY IS A CLASSPATH ARGUMENT, which is a fact
// about this repository rather than about the product: the sidecar ships as a JVM application and
// has no native launcher yet. The library's contract is unaffected - it spawns one process
// directly, with no shell between it and the child, so the pipe this process holds is the one the
// sidecar watches for parent death.
type sidecarCommand struct {
	path string
	args []string
}

// resolveSidecar assembles that command, or explains what is missing.
func resolveSidecar() (sidecarCommand, error) {
	java, err := javaLauncher()
	if err != nil {
		return sidecarCommand{}, err
	}
	classpath, err := sidecarClasspath()
	if err != nil {
		return sidecarCommand{}, err
	}
	return sidecarCommand{path: java, args: []string{"-cp", classpath, mainClass}}, nil
}

// javaLauncher resolves the JVM launcher to an absolute path, preferring JAVA_HOME.
//
// A PATH LOOKUP IS ACCEPTABLE HERE AND NOWHERE IN THE LIBRARY. The library refuses a relative or
// PATH-resolved SidecarPath because it hands that binary the user's Kafka credentials; this demo is
// choosing a JVM to run a proxy it also built, against a broker it also seeded, and the same
// reasoning that lets internal/harness look java up applies. What the library receives is still
// absolute, which is the property that matters to it.
func javaLauncher() (string, error) {
	if home := os.Getenv("JAVA_HOME"); home != "" {
		candidate := filepath.Join(home, "bin", "java")
		if _, err := os.Stat(candidate); err == nil {
			return candidate, nil
		}
	}
	found, err := exec.LookPath("java")
	if err != nil {
		return "", fmt.Errorf("the demo found no JVM to run the sidecar with - set JAVA_HOME, or "+
			"run the demo in its container (demo/run.sh --docker): %w", err)
	}
	return filepath.Abs(found)
}

// sidecarClasspath reads the classpath file the entry point wrote.
//
// It FAILS rather than guessing. A demo that silently ran a differently-built sidecar would report
// throughput for an engine nobody asked for, and the message names the command that produces the
// file rather than leaving the reader to find it.
func sidecarClasspath() (string, error) {
	path := strings.TrimSpace(os.Getenv(sidecarClasspathEnv))
	if path == "" {
		return "", errors.New(sidecarClasspathEnv + " is not set, so the demo does not know where " +
			"the sidecar is - start the demo through demo/run.sh, which builds it and sets this")
	}
	raw, err := os.ReadFile(path)
	if err != nil {
		return "", fmt.Errorf("reading the sidecar classpath from %s: %w", path, err)
	}
	classpath := strings.TrimSpace(string(raw))
	if classpath == "" {
		return "", fmt.Errorf("the sidecar classpath at %s is empty", path)
	}
	return classpath, nil
}
